### PR TITLE
Another attempt at fixing frame injection in the dispatcher tracing path

### PR DESCRIPTION
**Pull request by asodeur**
This is a slight variation of #3638 which is crashing for me (py3.6 & py3.7 on Windows 7).

Original issue is #3229

### CODE REVIEW

**Comment by esc** [comment]
@asodeur thanks for submitting this, we appreciate your efforts to improve Numba. I've labeled this as 'In Progress' for now as there are still some CI failures. Also, I noticed there seem to be a number of spurious commits in this PR and the commit-history will have to be fixed before this PR can be accepted. Do feel free to ping me in case you need some assistance with the appropriate `git` commands.

**Comment by stuartarchibald** [comment]
@asodeur open/close to get a new CI build, looks like you patched in the C89 stuff RE https://gitter.im/numba/numba?at=5dbac47ea3f0b17849bf6aa8 ?

File: numba/_dispatcher.c [hunk 1]
```diff
@@ -325,17 +328,21 @@ call_cfunc(DispatcherObject *self, PyObject *cfunc, PyObject *args, PyObject *kw
...
         }
         /* Populate the 'fast locals' in `frame` */
-        Py_XDECREF(frame->f_locals);
-        frame->f_locals = locals;
-        Py_XINCREF(frame->f_locals);
         PyFrame_LocalsToFast(frame, 0);
         tstate->frame = frame;
         C_TRACE(result, fn(PyCFunction_GET_SELF(cfunc), args, kws));
+        /* write changes back to locals? */
+        PyFrame_FastToLocals(frame);
```
Comment (stuartarchibald) [inline numba/_dispatcher.c]: I think this does something more than the manual attempt at resetting https://github.com/numba/numba/pull/3638/files#diff-4a1cc89a13f0d44d41dfa7e993591bffR336-R337 which might be why this seems to work?

**Comment by stuartarchibald** [comment]
TODO: We need to work out how to test this!

**Comment by asodeur** [comment]
Added a profiling testcase based on `numba.tests.npyufunc.test_dufunc.TestDUFunc.test_npm_call` (which happens to be the first segfaulting testcase when profiling) which is passing only after this fix is applied.This is not strictly a proof the fix will work under all circumstances but at least could be easily extended to include any other testcase segfaulting under the profiler.

**Comment by stuartarchibald** [comment]
> Added a profiling testcase based on `numba.tests.npyufunc.test_dufunc.TestDUFunc.test_npm_call` (which happens to be the first segfaulting testcase when profiling) which is passing only after this fix is applied.This is not strictly a proof the fix will work under all circumstances but at least could be easily extended to include any other testcase segfaulting under the profiler.

Thanks for adding this. I'm of the view that I'd still really like to understand precisely why this fixes it and the other 3 or 4 minor variants we tried didn't. However, practically, this seems to prevent the segfault, is now tested and demonstrating this as the case, and given that this is on a non-default code path I think we should go with it and wait for feedback/issues. Thoughts?

**Comment by stuartarchibald** [comment]
@sklam are you happy with the above too?

**Comment by stuartarchibald** [comment]
Farm build ID `numba_smoketest_cpu_25`.

**Comment by stuartarchibald** [comment]
Build farm passed.

**Comment by stuartarchibald** [comment]
Without this patch, the test added in this patch here https://github.com/numba/numba/pull/4476/files#diff-7d3763647d9a685ed9fb3b2c52e9ddf9R77-R95 and the test case here https://github.com/numba/numba/issues/3625#issue-393409644 both segfault with the backtrace starting from `python(+0x7ce97)[0x55b27b43ae97]` i.e. likely the same cause. With this patch, both samples work fine.

**Comment by stuartarchibald** [comment]
Reminder to close #3625 and #3229 when this is merged.

**Pull request merged**